File: Polimorfizam/Polimorfizam/ViewController.m
//
//  ViewController.m
//  Polimorfizam
//
//  Created by Djuro Alfirevic on 2/10/17.
//  Copyright © 2017 Djuro Alfirevic. All rights reserved.
//

#import "ViewController.h"
#import "ColorTableViewCell.h"

@interface ViewController() <UITableViewDataSource, UITableViewDelegate>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) NSMutableArray *itemsArray;
@end

@implementation ViewController

#pragma mark - Private API

- (void)configureArray {
    self.itemsArray = [[NSMutableArray alloc] init];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"GreenCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"BlueCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"YellowCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"BlackCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"ColorCell"];
    [self.itemsArray addObject:@"BlackCell"];

    [self.tableView reloadData];
}

#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];

    self.tableView.tableFooterView = [[UIView alloc] init];
    [self configureArray];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.itemsArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *identifier = self.itemsArray[indexPath.row];
    ColorTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    ColorTableViewCell *cell = (ColorTableViewCell *)[self tableView:self.tableView cellForRowAtIndexPath:indexPath];

    // [tableView cellForRowAtIndexPath:indexPath] -> Segue
    // [self tableView:self.tableView cellForRowAtIndexPath:indexPath];
    return [cell height];
}

@end

File: OMDB/OMDB/ViewController.swift
//
//  ViewController.swift
//  OMDB
//
//  Created by Djuro Alfirevic on 4/19/17.
//  Copyright © 2017 Sistemi. All rights reserved.
//

import UIKit

let kBaseURL = "http://www.omdbapi.com/?t="

class ViewController: UIViewController {

    // MARK: - Properties
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var searchBar: UISearchBar!
    var items = [String]()

    // MARK: - Private API
    fileprivate func search(_ term: String) {
        // Grab JSON from OMDB database.
        DispatchQueue.global(qos: .userInitiated).async {
            if let url = URL(string: "\(kBaseURL)\(term)") {
                if let data = try? Data(contentsOf: url) {
                    DispatchQueue.main.async { [unowned self] in
                        // Parsing JSON.
                        do {
                            let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]

                            if let json = json {
                                self.searchBar.text = ""

                                self.openMovie(Movie(json: json))
                            }
                        } catch {
                            print("Error: \(error.localizedDescription)")
                        }
                    }
                }
            }
        }
    }

    fileprivate func openMovie(_ movie: Movie) {
        let movieViewController = UIStoryboard(name: "Main", bundle: Bundle.main).instantiateViewController(withIdentifier: "MovieViewController") as! MovieViewController
        movieViewController.movie = movie
        navigationController?.pushViewController(movieViewController, animated: true)
    }

    // MARK: - View lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.tableFooterView = UIView()
    }

}

extension ViewController: UITableViewDataSource, UITableViewDelegate {

    // MARK: - UITableViewDataSource
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell")!

        cell.textLabel?.text = items[indexPath.row]

        return cell
    }

    // MARK: - UITableViewDelegate
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        search(items[indexPath.row])
    }

}

extension ViewController: UISearchBarDelegate {

    // MARK: - UISearchBarDelegate
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        if let text = searchBar.text {
            if !text.isEmpty {
                // Appen term to items.
                items.append(text)
                items.sort()

                tableView.reloadData()

                search(text)
            }
        }
    }


}
